### PR TITLE
Add APIs to show the progress of policyfile code rollouts

### PR DESCRIPTION
**Pull request by danielsdeleo**
### :nut_and_bolt: Description: What code changed, and why?

Add APIs to give statistics about the number of nodes having rolled out recent revisions of policies.

This work is done ahead of having a UI mockup, so it's a little bit speculative and probably will need a bit of updating to match the needs of the UI. That said, here's what it does:

* Queries the (pg) database for the 5 most recent rollouts per each node segment. Node segment is the policy name, policy group, and policy domain (Chef Server, for now) URL that groups nodes according to the code assigned to them.
* Queries the node-state database (in elastic) to aggregate node-state statistics, grouped by node segment plus policy revision. This gives us the number of nodes in each segment, the number of successes and failures for each policy revision, and the count of which revisions were applied on the most recent Chef Run by nodes within the segment.
* Merges the results of these queries into a single API response containing both detailed rollout information along with statistics about the Chef Client runs belonging to these rollouts.

### :chains: Related Resources

#3688 

### :athletic_shoe: How to Build and Test the Change

* Build config mgmt service and automate-gateway
* start all services (a subset probably works but I just ran them all)
* run `config_mgmt_list_rollouts_progress` to list the rollouts (this is a wrapper around `curl -fsS --insecure -H "api-token: ${api_token}" "https://localhost/api/beta/cfgmgmt/rollouts/progress_by_node_segment"` ) -- the response should be empty
* run `config_mgmt_populate_rollouts` to make sample data
* run `config_mgmt_list_rollouts_progress` to see the data.

### CODE REVIEW

File: components/config-mgmt-service/integration_test/rollout_progress_test.go [hunk 1]
```diff
@@ -0,0 +1,517 @@
+package integration_test
+
+/*
+
+TODO LIST:
+* fix the mark missing so we retain the prior status in the database
+* make sure crap like trailing `/` on URLs, http vs. https doesn't affect matching of rollouts
+* Use "errored" instead of failed to describe CCRs that had an error
```
Comment (danielsdeleo) [inline components/config-mgmt-service/integration_test/rollout_progress_test.go:8]: Also: add APIs to detail the known node segments (at least) so users can troubleshoot if their rollouts and CCRs aren't matching up.

File: components/config-mgmt-service/integration_test/rollout_progress_test.go [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+	// 	cleanup(t)
+	// })
+
+	// TODO
+	// t.Run("with variations in the Chef Server URL", func(t *testing.T) {
+	// 	// We should do the right thing for:
+	// 	// * trailing slash
+	// 	// * http URL given to the rollout
+	// 	cleanup(t)
+	// })
```
Comment (danielsdeleo) [inline components/config-mgmt-service/integration_test/rollout_progress_test.go:465]: Delete these TODOs if I'm gonna move them to a separate PR.

File: components/config-mgmt-service/backend/client.go [hunk 3]
```diff
@@ -201,3 +202,28 @@ type ChefErrorCount struct {
...
+	BySegment map[NodeSegment]*NodeSegmentRevisionsStatus
+}
+
+type NodeSegmentRevisionsStatus struct {
+	NodeSegment
+	NodesInSegment   int32
+	ByPolicyRevision map[string]*PolicyRevisionNodeStatus
+}
+
+// FIXME: get rid of "missing" here, we don't care (seems like it will need a kinda deep change to ingest/storage)
```
Comment (danielsdeleo) [inline components/config-mgmt-service/backend/client.go:222]: Remove TODO if it will be done in a separate PR

File: api/external/cfgmgmt/response/rollouts.proto [hunk 4]
```diff
@@ -85,3 +85,56 @@ message Rollouts {
...
+}
+
+message CurrentRolloutProgress {
+  // Rollout is the full rollout object, but we can change this to be a subset only.
+  Rollout rollout = 1;
+  // Nodes that have run the code being rolled out thus far
+  int32 node_count = 2;
+  // I'm assuming it's easy to get the status when we get the counts.
+  int32 latest_run_successful_count = 3;
+  int32 latest_run_failed_count = 4;
```
Comment (danielsdeleo) [inline api/external/cfgmgmt/response/rollouts.proto:124]: Use "errored" instead of "failed" in general

File: components/config-mgmt-service/backend/elastic/node_segment_rollouts.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+			"page_request_count":   pctx.requestCount,
+			"results_so_far_count": len(pctx.results),
+		}).Error("aborting aggregation request")
+		return newErr
+	}
+	pctx.HandleResults(aggData)
+	return nil
+}
+
+type paginationContext struct {
```
Comment (lancewf) [inline components/config-mgmt-service/backend/elastic/node_segment_rollouts.go:180]: This is a nice tool.

File: components/config-mgmt-service/backend/postgres/node_segments.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+const normalizedURLScheme = "https"
+
+const mostRecentRolloutsForSegments = `
+  SELECT rollouts_by_node_segment.*
+    FROM (
+         SELECT r.*, rank()
+				   OVER (
+					        PARTITION BY policy_node_group, policy_name, policy_domain_url
+									    ORDER BY start_time DESC, id
+								) AS order_index
```
Comment (lancewf) [inline components/config-mgmt-service/backend/postgres/node_segments.go:28]: :exploding_head: I am going to save this for later.
